Test/diffs/difftester.py

```diff
@@ -0,0 +1,955 @@
+#!/usr/bin/env python3
+r"""A differential tester for Dafny.
```

Comment:
Why is Python the right language for this? I don't expect Dafny engineers to be experts at any language other than the main language Dafny is written in, so I think not having it in C# severely decreases the maintainability. That isn't a deal breaker but there needs to be a good reason.

Reply:
There is no good reason, except that it made it easier to prototype.  I'm happy to port it to C# if we find that better.

Reply:
I think having it in C# is much more maintainable. I worked on a product defined in many languages before and there were significant differences in code quality between the most used language and others.

Reply:
I agree that c# is more maintainable. However, it's possible to immediately run a python file, whereas it's not with a c#.
If we just use this script for smoking tests down the road or like a tool in our tool belt, then we are fine keeping it in Python. If we plan to integrate it in the CI or to automate the process, then C# might be a better option.

Reply:
> However, it's possible to immediately run a python file, whereas it's not with a c#.

Can you elaborate? What's the difference in UX between `python file.py` and `dotnet run file.csproj` ?

Reply:
That's the point, you have to run a `dotnet run file.csproj`, you can't run `dotnet run file.cs`, you have to create one extra file and run it, with three commands. Moreover, you need to start a dotnet virtual machine for that.
Whereas in python, with just a comment at the beginning of the file, you can run it like an executable like `./file.py`

Reply:
Seems like you can do the same with .NET but then using `dotnet-script program.csx`, after running `dotnet tool install -g dotnet-script`

> Moreover, you need to start a dotnet virtual machine for that.

Does it matter whether you start a .NET or a Python one runtime?

Reply:
I know it matters for Java (a JVM takes time to start), I'm not sure for .NET

Still, two commands, and moreover you need to install a tool(?)

Reply:
The `dotnet-script` tool would be installed by placing it in the `dotnet-tools.json` file we already have. For Python, ideally we also don't assume anything about what Python version the user has installed as a default. Maybe we would place Python installation instructions in a README or add something to our makefile. I'm not sure what's idiomatic there.

I'm a bit confused by this discussion. Doesn't the benefit of having the repository be in a single language far outweigh these differences in how to run the script?

Reply:
Sorry for the confusion. Thanks for digging deeper.
Having the repository to be (more) in a single language outweighs a little - in my opinion - these differences in *how to run the script*. If we had the c# script, we could always make an extra shell wrapper around it to run it, so that wouldn't be a problem.

Now, the question of this thread is: do we want @cpitclaudel  to actually rewrite this in C#? I argue that we don't need to spend time doing that. This script is not integrated with the Dafny source files, it's only running processes. We would gain nothing porting a program of this size to a typed language. This script is not going to evolve substantially, is very likely to not going to have unit tests.

Each language has its own idiomatic use. Python has been designed for scripting and quick prototyping, file and process handling and so on, C# has been designed for building applications and robust shared libraries. Hence, I understand the current choice, although I would have never pushed anyone to use Python instead of C# if they preferred to start their project with the latter.

Reply:
@MikaelMayer So, disregarding what language it's currently written in, what language do you prefer it to be written in, or are you ambivalent about that?

Reply:
Disregarding the language our codebase is currently written in, I would prefer it to be written in Python.